mysticeti-core/src/consensus/universal_committer.rs

```diff
@@ -51,28 +61,34 @@ impl UniversalCommitter {
                 tracing::debug!("Outcome of direct rule: {status}");
 
                 // If we can't directly decide the leader, try to indirectly decide it.
-                if !status.is_decided() {
-                    status = committer.try_indirect_decide(leader, round, leaders.iter());
-                    self.update_metrics(&status, false);
+                if status.is_decided() {
+                    leaders.push_front((status.clone(), true));
+                } else {
+                    status = committer.try_indirect_decide(
+                        leader,
+                        round,
+                        leaders.iter().map(|(x, _)| x),
+                    );
+                    leaders.push_front((status.clone(), false));
                     tracing::debug!("Outcome of indirect rule: {status}");
                 }
-
-                leaders.push_front(status);
             }
         }
 
         // The decided sequence is the longest prefix of decided leaders.
         leaders
             .into_iter()
-            // Skip all leaders before the last decided round.
-            .skip_while(|x| (x.round(), x.authority()) != last_decided_round_authority)
-            // Skip the last decided leader.
-            .skip(1)
             // Filter out all the genesis.
-            .filter(|x| x.round() > 0)
+            .filter(|(x, _)| x.round() > 0)
             // Stop the sequence upon encountering an undecided leader.
-            .take_while(|x| x.is_decided())
-            .inspect(|x| tracing::debug!("Decided {x}"))
+            .take_while(|(x, _)| x.is_decided())
+            // We want to report metrics at this point to ensure that the decisions are reported only once
+            // hence we increase our accuracy
+            .inspect(|(x, direct_decided)| {
+                self.update_metrics(x, *direct_decided);
+                tracing::debug!("Decided {x}");
+            })
+            .map(|(x, _)| x)
             .collect()
```

Comment:
I am all for changing this logic (it was quite messy) but let's make sure we test the hell out of it. I got it wrong more than once in the past.

Reply:
For sure I can test more that, but for initial feedback I was mostly based on the existing unit tests